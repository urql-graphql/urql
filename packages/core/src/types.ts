import { TypedDocumentNode } from '@graphql-typed-document-node/core';
import { GraphQLError, DocumentNode } from 'graphql';
import { Source } from 'wonka';
import { Client } from './client';
import { CombinedError } from './utils/error';

export type ExecutionResult =
  | {
      errors?:
        | Array<Partial<GraphQLError> | string | Error>
        | readonly GraphQLError[];
      data?: null | Record<string, any>;
      extensions?: Record<string, any>;
      hasNext?: boolean;
    }
  | {
      errors?:
        | Array<Partial<GraphQLError> | string | Error>
        | readonly GraphQLError[];
      data: any;
      path: (string | number)[];
      hasNext?: boolean;
    };

export type PromisifiedSource<T = any> = Source<T> & {
  toPromise: () => Promise<T>;
};

/** The type of GraphQL operation being executed. */
export type OperationType = 'subscription' | 'query' | 'mutation' | 'teardown';

/** The strategy that is used to request results from network and/or the cache. */
export type RequestPolicy =
  | 'cache-first'
  | 'cache-only'
  | 'network-only'
  | 'cache-and-network';

/** How the operation has */
export type CacheOutcome = 'miss' | 'partial' | 'hit';

/** A default type for variables */
export type AnyVariables = { [prop: string]: any } | void | undefined;

/** A Graphql query, mutation, or subscription. */
export interface GraphQLRequest<
  Data = any,
  Variables extends AnyVariables = AnyVariables
> {
  /** Unique identifier of the request. */
  key: number;
  query: DocumentNode | TypedDocumentNode<Data, Variables>;
  variables: Variables;
}

/** Metadata that is only available in development for devtools. */
export interface OperationDebugMeta {
  source?: string;
  cacheOutcome?: CacheOutcome;
  networkLatency?: number;
  startTime?: number;
}

/** A unique marker that marks the operation's identity when multiple mutation operations with identical keys are issued. */
export type OperationInstance = number & { readonly _opaque: unique symbol };

/** Additional metadata passed to [exchange]{@link Exchange} functions. */
export interface OperationContext {
  [key: string]: any;
  readonly _instance?: OperationInstance | undefined;
  additionalTypenames?: string[];
  fetch?: typeof fetch;
  fetchOptions?: RequestInit | (() => RequestInit);
  requestPolicy: RequestPolicy;
  url: string;
  meta?: OperationDebugMeta;
  suspense?: boolean;
  /** Instructs fetch exchanges to use a GET request.
   * When true or 'within-url-limit' is passed, GET won't be used when the resulting URL exceeds a length of 2048. */
  preferGetMethod?: boolean | 'force' | 'within-url-limit';
}

/** A [query]{@link Query} or [mutation]{@link Mutation} with additional metadata for use during transmission. */
export interface Operation<
  Data = any,
  Variables extends AnyVariables = AnyVariables
> extends GraphQLRequest<Data, Variables> {
  readonly kind: OperationType;
  context: OperationContext;
}

/** Resulting data from an [operation]{@link Operation}. */
export interface OperationResult<
  Data = any,
  Variables extends AnyVariables = AnyVariables
> {
  /** The [operation]{@link Operation} which has been executed. */
  operation: Operation<Data, Variables>;
  /** The data returned from the Graphql server. */
  data?: Data;
  /** Any errors resulting from the operation. */
  error?: CombinedError;
  /** Optional extensions return by the Graphql server. */
  extensions?: Record<string, any>;
  /** Optional stale flag added by exchanges that return stale results. */
  stale?: boolean;
  /** Optional hasNext flag indicating deferred/streamed results are following. */
  hasNext?: boolean;
}

/** Input parameters for to an Exchange factory function. */
export interface ExchangeInput {
  client: Client;
  forward: ExchangeIO;
  dispatchDebug: <T extends keyof DebugEventTypes | string>(
    t: DebugEventArg<T>
  ) => void;
}

/** Function responsible for listening for streamed [operations]{@link Operation}. */
export type Exchange = (input: ExchangeInput) => ExchangeIO;

/** Function responsible for receiving an observable [operation]{@link Operation} and returning a [result]{@link OperationResult}. */
export type ExchangeIO = (ops$: Source<Operation>) => Source<OperationResult>;

/** Debug event types (interfaced for declaration merging). */
export interface DebugEventTypes {
  // Cache exchange
  cacheHit: { value: any };
  cacheInvalidation: {
    typenames: string[];
    response: OperationResult;
  };
  // Fetch exchange
  fetchRequest: {
    url: string;
    fetchOptions: RequestInit;
  };
  fetchSuccess: {
    url: string;
    fetchOptions: RequestInit;
    value: object;
  };
  fetchError: {
    url: string;
    fetchOptions: RequestInit;
    value: Error;
  };
  // Retry exchange
  retryRetrying: {
    retryCount: number;
  };
}

export type DebugEventArg<T extends keyof DebugEventTypes | string> = {
  type: T;
  message: string;
  operation: Operation;
} & (T extends keyof DebugEventTypes
  ? { data: DebugEventTypes[T] }
  : { data?: any });

export type DebugEvent<
  T extends keyof DebugEventTypes | string = string
> = DebugEventArg<T> & {
  timestamp: number;
  source: string;
};
