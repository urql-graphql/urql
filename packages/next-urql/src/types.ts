import { GraphQLError } from 'graphql';
import { ClientOptions, Exchange, Client } from 'urql';
import { NextPageContext } from 'next';
import { AppContext } from 'next/app';

export interface NextUrqlPageContext extends NextPageContext {
  urqlClient: Client;
}

export interface NextUrqlAppContext extends AppContext {
  urqlClient: Client;
}

export type NextUrqlContext = NextUrqlPageContext | NextUrqlAppContext;

export type NextUrqlClientConfig = (
  ssrExchange: SSRExchange,
  ctx?: NextPageContext
) => ClientOptions;

export interface WithUrqlState {
  urqlState?: SSRData;
}

export interface WithUrqlClient {
  urqlClient?: Client;
}

export interface WithUrqlProps extends WithUrqlClient, WithUrqlState {
  resetUrqlClient?: () => void;
  pageProps: any;
  [key: string]: any;
}

export interface SerializedResult {
  data?: any;
  error?: {
    graphQLErrors: Array<Partial<GraphQLError> | string>;
    networkError?: string;
  };
}

export interface SSRData {
  [key: string]: SerializedResult;
}

export interface SSRExchange extends Exchange {
  /** Rehydrates cached data */
  restoreData(data: SSRData): void;
  /** Extracts cached data */
  extractData(): SSRData;
}

export interface WithUrqlClientOptions {
  ssr?: boolean;
  neverSuspend?: boolean;
  staleWhileRevalidate?: boolean;
}
