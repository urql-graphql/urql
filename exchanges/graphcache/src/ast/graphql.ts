import type * as GraphQL from 'graphql';

type OrNever<T> = 0 extends 1 & T ? never : T;

export type IntrospectionQuery =
  | {
      readonly __schema: {
        queryType: { name: string; kind?: any };
        mutationType?: { name: string; kind?: any } | null;
        subscriptionType?: { name: string; kind?: any } | null;
        types?: readonly IntrospectionType[];
      };
    }
  | OrNever<GraphQL.IntrospectionQuery>;

export type IntrospectionTypeRef =
  | {
      readonly kind:
        | 'SCALAR'
        | 'OBJECT'
        | 'INTERFACE'
        | 'ENUM'
        | 'UNION'
        | 'INPUT_OBJECT';
      readonly name?: string;
      readonly ofType?: IntrospectionTypeRef;
    }
  | OrNever<GraphQL.IntrospectionTypeRef>;

export type IntrospectionInputTypeRef =
  | {
      readonly kind: 'SCALAR' | 'ENUM' | 'INPUT_OBJECT';
      readonly name?: string;
      readonly ofType?: IntrospectionInputTypeRef;
    }
  | OrNever<GraphQL.IntrospectionInputTypeRef>;

export type IntrospectionInputValue =
  | {
      readonly name: string;
      readonly description?: string | null;
      readonly defaultValue?: string | null;
      readonly type: IntrospectionInputTypeRef;
    }
  | OrNever<GraphQL.IntrospectionInputValue>;

export type IntrospectionType =
  | {
      readonly kind: string;
      readonly name: string;
      readonly fields?: readonly any[];
      readonly interfaces?: readonly any[];
      readonly possibleTypes?: readonly any[];
    }
  | OrNever<GraphQL.IntrospectionType>;
