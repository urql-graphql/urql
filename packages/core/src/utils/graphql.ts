import type * as GraphQLWeb from '@0no-co/graphql.web';
import type * as GraphQL from 'graphql';

type OrNever<T> = 0 extends 1 & T ? never : T;

export type GraphQLError =
  | GraphQLWeb.GraphQLError
  | OrNever<GraphQL.GraphQLError>;

export type DocumentNode =
  | GraphQLWeb.DocumentNode
  | OrNever<GraphQL.DocumentNode>;

export type DefinitionNode =
  | GraphQLWeb.DefinitionNode
  | OrNever<GraphQL.DefinitionNode>;
