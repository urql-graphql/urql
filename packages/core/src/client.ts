/* eslint-disable @typescript-eslint/no-use-before-define */

import {
  filter,
  make,
  makeSubject,
  onEnd,
  onPush,
  onStart,
  pipe,
  share,
  Source,
  take,
  takeUntil,
  publish,
  subscribe,
  switchMap,
  fromValue,
  merge,
  map,
  Subscription,
} from 'wonka';

import { TypedDocumentNode } from '@graphql-typed-document-node/core';
import { DocumentNode } from 'graphql';

import { composeExchanges, defaultExchanges } from './exchanges';
import { fallbackExchange } from './exchanges/fallback';

import {
  Exchange,
  ExchangeInput,
  GraphQLRequest,
  Operation,
  OperationContext,
  OperationResult,
  OperationType,
  RequestPolicy,
  PromisifiedSource,
  DebugEvent,
} from './types';

import {
  createRequest,
  withPromise,
  maskTypename,
  noop,
  makeOperation,
  getOperationType,
} from './utils';

/** Options for configuring the URQL [client]{@link Client}. */
export interface ClientOptions {
  /** Target endpoint URL such as `https://my-target:8080/graphql`. */
  url: string;
  /** Any additional options to pass to fetch. */
  fetchOptions?: RequestInit | (() => RequestInit);
  /** An alternative fetch implementation. */
  fetch?: typeof fetch;
  /** An ordered array of Exchanges. */
  exchanges?: Exchange[];
  /** Activates support for Suspense. */
  suspense?: boolean;
  /** The default request policy for requests. */
  requestPolicy?: RequestPolicy;
  /** Use HTTP GET for queries. */
  preferGetMethod?: boolean;
  /** Mask __typename from results. */
  maskTypename?: boolean;
}

export interface Client {
  new (options: ClientOptions): Client;

  operations$: Source<Operation>;

  /** Start an operation from an exchange */
  reexecuteOperation: (operation: Operation) => void;
  /** Event target for monitoring, e.g. for @urql/devtools */
  subscribeToDebugTarget?: (onEvent: (e: DebugEvent) => void) => Subscription;

  // These are variables derived from ClientOptions
  url: string;
  fetch?: typeof fetch;
  fetchOptions?: RequestInit | (() => RequestInit);
  suspense: boolean;
  requestPolicy: RequestPolicy;
  preferGetMethod: boolean;
  maskTypename: boolean;

  createOperationContext(
    opts?: Partial<OperationContext> | undefined
  ): OperationContext;

  createRequestOperation<Data = any, Variables = object>(
    kind: OperationType,
    request: GraphQLRequest<Data, Variables>,
    opts?: Partial<OperationContext> | undefined
  ): Operation<Data, Variables>;

  /** Executes an Operation by sending it through the exchange pipeline It returns an observable that emits all related exchange results and keeps track of this observable's subscribers. A teardown signal will be emitted when no subscribers are listening anymore. */
  executeRequestOperation<Data = any, Variables = object>(
    operation: Operation<Data, Variables>
  ): Source<OperationResult<Data, Variables>>;

  query<Data = any, Variables extends object = {}>(
    query: DocumentNode | TypedDocumentNode<Data, Variables> | string,
    variables?: Variables,
    context?: Partial<OperationContext>
  ): PromisifiedSource<OperationResult<Data, Variables>>;

  readQuery<Data = any, Variables extends object = {}>(
    query: DocumentNode | TypedDocumentNode<Data, Variables> | string,
    variables?: Variables,
    context?: Partial<OperationContext>
  ): OperationResult<Data, Variables> | null;

  executeQuery<Data = any, Variables = object>(
    query: GraphQLRequest<Data, Variables>,
    opts?: Partial<OperationContext> | undefined
  ): Source<OperationResult<Data, Variables>>;

  subscription<Data = any, Variables extends object = {}>(
    query: DocumentNode | TypedDocumentNode<Data, Variables> | string,
    variables?: Variables,
    context?: Partial<OperationContext>
  ): Source<OperationResult<Data, Variables>>;

  executeSubscription<Data = any, Variables = object>(
    query: GraphQLRequest<Data, Variables>,
    opts?: Partial<OperationContext> | undefined
  ): Source<OperationResult<Data, Variables>>;

  mutation<Data = any, Variables extends object = {}>(
    query: DocumentNode | TypedDocumentNode<Data, Variables> | string,
    variables?: Variables,
    context?: Partial<OperationContext>
  ): PromisifiedSource<OperationResult<Data, Variables>>;

  executeMutation<Data = any, Variables = object>(
    query: GraphQLRequest<Data, Variables>,
    opts?: Partial<OperationContext> | undefined
  ): Source<OperationResult<Data, Variables>>;
}

export const Client: new (opts: ClientOptions) => Client = function Client(
  this: Client | {},
  opts: ClientOptions
) {
  if (process.env.NODE_ENV !== 'production' && !opts.url) {
    throw new Error('You are creating an urql-client without a url.');
  }

  const replays = new Map<number, OperationResult>();
  const active: Map<number, Source<OperationResult>> = new Map();
  const queue: Operation[] = [];

  // This subject forms the input of operations; executeOperation may be
  // called to dispatch a new operation on the subject
  const { source: operations$, next: nextOperation } = makeSubject<Operation>();

  // We define a queued dispatcher on the subject, which empties the queue when it's
  // activated to allow `reexecuteOperation` to be trampoline-scheduled
  let isOperationBatchActive = false;
  function dispatchOperation(operation?: Operation | void) {
    isOperationBatchActive = true;
    if (operation) nextOperation(operation);
    while ((operation = queue.shift())) nextOperation(operation);
    isOperationBatchActive = false;
  }

  /** Defines how result streams are created */
  const makeResultSource = (operation: Operation) => {
    let result$ = pipe(
      results$,
      filter(
        (res: OperationResult) =>
          res.operation.kind === operation.kind &&
          res.operation.key === operation.key
      )
    );

    // Mask typename properties if the option for it is turned on
    if (client.maskTypename) {
      result$ = pipe(
        result$,
        map(res => ({ ...res, data: maskTypename(res.data) }))
      );
    }

    // A mutation is always limited to just a single result and is never shared
    if (operation.kind === 'mutation') {
      return pipe(
        result$,
        onStart(() => dispatchOperation(operation)),
        take(1)
      );
    }

    const source = pipe(
      result$,
      // End the results stream when an active teardown event is sent
      takeUntil(
        pipe(
          operations$,
          filter(op => op.kind === 'teardown' && op.key === operation.key)
        )
      ),
      switchMap(result => {
        if (operation.kind !== 'query' || result.stale) {
          return fromValue(result);
        }

        return merge([
          fromValue(result),
          // Mark a result as stale when a new operation is sent for it
          pipe(
            operations$,
            filter(
              op =>
                op.kind === 'query' &&
                op.key === operation.key &&
                op.context.requestPolicy !== 'cache-only'
            ),
            take(1),
            map(() => ({ ...result, stale: true }))
          ),
        ]);
      }),
      onPush(result => {
        replays.set(operation.key, result);
      }),
      onStart(() => {
        active.set(operation.key, source);
      }),
      onEnd(() => {
        // Delete the active operation handle
        replays.delete(operation.key);
        active.delete(operation.key);
        // Delete all queued up operations of the same key on end
        for (let i = queue.length - 1; i >= 0; i--)
          if (queue[i].key === operation.key) queue.splice(i, 1);
        // Dispatch a teardown signal for the stopped operation
        dispatchOperation(
          makeOperation('teardown', operation, operation.context)
        );
      }),
      share
    );

    return source;
  };

  const instance: Client =
    this instanceof Client ? this : Object.create(Client.prototype);
  const client: Client = Object.assign(instance, {
    url: opts.url,
    fetchOptions: opts.fetchOptions,
    fetch: opts.fetch,
    suspense: !!opts.suspense,
    requestPolicy: opts.requestPolicy || 'cache-first',
    preferGetMethod: !!opts.preferGetMethod,
    maskTypename: !!opts.maskTypename,

    operations$,

    reexecuteOperation(operation: Operation) {
      // Reexecute operation only if any subscribers are still subscribed to the
      // operation's exchange results
      if (operation.kind === 'mutation' || active.has(operation.key)) {
        queue.push(operation);
        if (!isOperationBatchActive) {
          Promise.resolve().then(dispatchOperation);
        }
      }
    },

    createOperationContext(opts) {
      if (!opts) opts = {};

      return {
        url: client.url,
        fetchOptions: client.fetchOptions,
        fetch: client.fetch,
        preferGetMethod: client.preferGetMethod,
        ...opts,
        suspense: opts.suspense || (opts.suspense !== false && client.suspense),
        requestPolicy: opts.requestPolicy || client.requestPolicy,
      };
    },

    createRequestOperation(kind, request, opts) {
      const requestOperationType = getOperationType(request.query);
      if (
        process.env.NODE_ENV !== 'production' &&
        kind !== 'teardown' &&
        requestOperationType !== kind
      ) {
        throw new Error(
          `Expected operation of type "${kind}" but found "${requestOperationType}"`
        );
      }
      return makeOperation(kind, request, client.createOperationContext(opts));
    },

    executeRequestOperation(operation) {
      if (operation.kind === 'mutation') {
        return makeResultSource(operation);
      }

      const source = active.get(operation.key) || makeResultSource(operation);

      const isNetworkOperation =
        operation.context.requestPolicy === 'cache-and-network' ||
        operation.context.requestPolicy === 'network-only';

      return make(observer => {
        return pipe(
          source,
          onStart(() => {
            const prevReplay = replays.get(operation.key);

            if (operation.kind === 'subscription') {
              return dispatchOperation(operation);
            } else if (isNetworkOperation) {
              dispatchOperation(operation);
            }

            if (
              prevReplay != null &&
              prevReplay === replays.get(operation.key)
            ) {
              observer.next(
                isNetworkOperation ? { ...prevReplay, stale: true } : prevReplay
              );
            } else if (!isNetworkOperation) {
              dispatchOperation(operation);
            }
          }),
          onEnd(observer.complete),
          subscribe(observer.next)
        ).unsubscribe;
      });
    },

    executeQuery(query, opts) {
      const operation = client.createRequestOperation('query', query, opts);
      return client.executeRequestOperation(operation);
    },

    executeSubscription(query, opts) {
      const operation = client.createRequestOperation(
        'subscription',
        query,
        opts
      );
      return client.executeRequestOperation(operation);
    },

    executeMutation(query, opts) {
      const operation = client.createRequestOperation('mutation', query, opts);
      return client.executeRequestOperation(operation);
    },

    query(query, variables, context) {
      if (!context || typeof context.suspense !== 'boolean') {
        context = { ...context, suspense: false };
      }

      return withPromise(
        client.executeQuery(createRequest(query, variables), context)
      );
    },

    readQuery(query, variables, context) {
      let result: OperationResult | null = null;

      pipe(
        client.query(query, variables, context),
        subscribe(res => {
          result = res;
        })
      ).unsubscribe();

      return result;
    },

    subscription(query, variables, context) {
      return client.executeSubscription(
        createRequest(query, variables),
        context
      );
    },

    mutation(query, variables, context) {
      return withPromise(
        client.executeMutation(createRequest(query, variables), context)
      );
    },
  } as Client);

  let dispatchDebug: ExchangeInput['dispatchDebug'] = noop;
  if (process.env.NODE_ENV !== 'production') {
    const { next, source } = makeSubject<DebugEvent>();
    client.subscribeToDebugTarget = (onEvent: (e: DebugEvent) => void) =>
      pipe(source, subscribe(onEvent));
    dispatchDebug = next as ExchangeInput['dispatchDebug'];
  }

  const exchanges =
    opts.exchanges !== undefined ? opts.exchanges : defaultExchanges;

  // All exchange are composed into a single one and are called using the constructed client
  // and the fallback exchange stream
  const composedExchange = composeExchanges(exchanges);

  // All exchanges receive inputs using which they can forward operations to the next exchange
  // and receive a stream of results in return, access the client, or dispatch debugging events
  // All operations then run through the Exchange IOs in a pipeline-like fashion
  const results$ = share(
    composedExchange({
      client,
      dispatchDebug,
      forward: fallbackExchange({ dispatchDebug }),
    })(operations$)
  );

  // Prevent the `results$` exchange pipeline from being closed by active
  // cancellations cascading up from components
  pipe(results$, publish);

  return client;
} as any;

export const createClient = (Client as any) as (opts: ClientOptions) => Client;
