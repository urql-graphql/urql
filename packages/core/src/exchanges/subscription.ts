import { print } from 'graphql';

import {
  filter,
  make,
  merge,
  mergeMap,
  pipe,
  share,
  Source,
  takeUntil,
} from 'wonka';

import { makeResult, makeErrorResult, makeOperation } from '../utils';

import {
  Exchange,
  ExecutionResult,
  Operation,
  OperationContext,
  OperationResult,
} from '../types';

export interface ObserverLike<T> {
  next: (value: T) => void;
  error: (err: any) => void;
  complete: () => void;
}

/** An abstract observable interface conforming to: https://github.com/tc39/proposal-observable */
export interface ObservableLike<T> {
  subscribe(
    observer: ObserverLike<T>
  ): {
    unsubscribe: () => void;
  };
}

export interface SubscriptionOperation {
  query: string;
  variables?: Record<string, unknown>;
  key: string;
  context: OperationContext;
}

export type SubscriptionForwarder = (
  operation: SubscriptionOperation
) => ObservableLike<ExecutionResult>;

/** This is called to create a subscription and needs to be hooked up to a transport client. */
export interface SubscriptionExchangeOpts {
  // This has been modelled to work with subscription-transport-ws
  // See: https://github.com/apollographql/subscriptions-transport-ws#requestoptions--observableexecutionresult-returns-observable-to-execute-the-operation
  forwardSubscription: SubscriptionForwarder;

  /** This flag may be turned on to allow your subscriptions-transport to handle all operation types */
  enableAllOperations?: boolean;
}

export const subscriptionExchange = ({
  forwardSubscription,
  enableAllOperations,
}: SubscriptionExchangeOpts): Exchange => ({ client, forward }) => {
  const createSubscriptionSource = (
    operation: Operation
  ): Source<OperationResult> => {
    // This excludes the query's name as a field although subscription-transport-ws does accept it since it's optional
    const observableish = forwardSubscription({
      key: operation.key.toString(36),
      query: print(operation.query),
      variables: operation.variables,
      context: { ...operation.context },
    });

    return make<OperationResult>(({ next, complete }) => {
      let isComplete = false;
      let sub;

      Promise.resolve().then(() => {
        if (isComplete) return;

        sub = observableish.subscribe({
          next: result => next(makeResult(operation, result)),
          error: err => next(makeErrorResult(operation, err)),
          complete: () => {
            if (!isComplete) {
              isComplete = true;
              if (operation.kind === 'subscription') {
                client.reexecuteOperation(
                  makeOperation('teardown', operation, operation.context)
                );
              }

              complete();
            }
          },
        });
      });

      return () => {
        isComplete = true;
        if (sub) sub.unsubscribe();
      };
    });
  };

  const isSubscriptionOperation = (operation: Operation): boolean => {
    const { kind } = operation;
    return (
      kind === 'subscription' ||
      (!!enableAllOperations && (kind === 'query' || kind === 'mutation'))
    );
  };

  return ops$ => {
    const sharedOps$ = share(ops$);
    const subscriptionResults$ = pipe(
      sharedOps$,
      filter(isSubscriptionOperation),
      mergeMap(operation => {
        const { key } = operation;
        const teardown$ = pipe(
          sharedOps$,
          filter(op => op.kind === 'teardown' && op.key === key)
        );

        return pipe(createSubscriptionSource(operation), takeUntil(teardown$));
      })
    );

    const forward$ = pipe(
      sharedOps$,
      filter(op => !isSubscriptionOperation(op)),
      forward
    );

    return merge([subscriptionResults$, forward$]);
  };
};
