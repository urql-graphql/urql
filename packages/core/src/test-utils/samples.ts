import { gql } from '../gql';

import {
  ExecutionResult,
  GraphQLRequest,
  Operation,
  OperationContext,
  OperationResult,
} from '../types';
import { makeOperation } from '../utils';

export const context: OperationContext = {
  fetchOptions: {
    method: 'POST',
  },
  requestPolicy: 'cache-first',
  url: 'http://localhost:3000/graphql',
};

export const queryGql: GraphQLRequest = {
  key: 2,
  query: gql`
    query getUser($name: String) {
      user(name: $name) {
        id
        firstName
        lastName
      }
    }
  `,
  variables: {
    name: 'Clara',
  },
};

export const mutationGql: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation AddUser($name: String) {
      addUser(name: $name) {
        name
      }
    }
  `,
  variables: {
    name: 'Clara',
  },
};

export const subscriptionGql: GraphQLRequest = {
  key: 4,
  query: gql`
    subscription subscribeToUser($user: String) {
      user(user: $user) {
        name
      }
    }
  `,
  variables: {
    user: 'colin',
  },
};

export const queryOperation: Operation = makeOperation(
  'query',
  {
    query: queryGql.query,
    variables: queryGql.variables,
    key: queryGql.key,
  },
  context
);

export const teardownOperation: Operation = makeOperation(
  'teardown',
  {
    query: queryOperation.query,
    variables: queryOperation.variables,
    key: queryOperation.key,
  },
  context
);

export const mutationOperation: Operation = makeOperation(
  'mutation',
  {
    query: mutationGql.query,
    variables: mutationGql.variables,
    key: mutationGql.key,
  },
  context
);

export const subscriptionOperation: Operation = makeOperation(
  'subscription',
  {
    query: subscriptionGql.query,
    variables: subscriptionGql.variables,
    key: subscriptionGql.key,
  },
  context
);

export const undefinedQueryResponse: OperationResult = {
  operation: queryOperation,
};

export const queryResponse: OperationResult = {
  operation: queryOperation,
  data: {
    user: {
      name: 'Clive',
    },
  },
};

export const mutationResponse: OperationResult = {
  operation: mutationOperation,
  data: {},
};

export const subscriptionResult: ExecutionResult = {
  data: {},
};
