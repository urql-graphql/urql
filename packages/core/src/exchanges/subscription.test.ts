import { print } from 'graphql';
import {
  empty,
  publish,
  fromValue,
  pipe,
  Source,
  take,
  toPromise,
} from 'wonka';
import { Client } from '../client';
import { subscriptionOperation, subscriptionResult } from '../test-utils';
import { OperationResult } from '../types';
import { subscriptionExchange, SubscriptionForwarder } from './subscription';

it('should return response data from forwardSubscription observable', async () => {
  const exchangeArgs = {
    dispatchDebug: jest.fn(),
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const unsubscribe = jest.fn();
  const forwardSubscription: SubscriptionForwarder = operation => {
    expect(operation.query).toBe(print(subscriptionOperation.query));
    expect(operation.variables).toBe(subscriptionOperation.variables);
    expect(operation.context).toEqual(subscriptionOperation.context);

    return {
      subscribe(observer) {
        Promise.resolve().then(() => {
          observer.next(subscriptionResult);
        });

        return { unsubscribe };
      },
    };
  };

  const data = await pipe(
    fromValue(subscriptionOperation),
    subscriptionExchange({ forwardSubscription })(exchangeArgs),
    take(1),
    toPromise
  );

  expect(data).toMatchSnapshot();
  expect(unsubscribe).toHaveBeenCalled();
});

it('should tear down the operation if the source subscription ends', async () => {
  const reexecuteOperation = jest.fn();
  const unsubscribe = jest.fn();

  const exchangeArgs = {
    dispatchDebug: jest.fn(),
    forward: () => empty as Source<OperationResult>,
    client: { reexecuteOperation: reexecuteOperation as any } as Client,
  };

  const forwardSubscription: SubscriptionForwarder = () => ({
    subscribe(observer) {
      observer.complete();
      return { unsubscribe };
    },
  });

  pipe(
    fromValue(subscriptionOperation),
    subscriptionExchange({ forwardSubscription })(exchangeArgs),
    publish
  );

  await Promise.resolve();

  expect(unsubscribe).not.toHaveBeenCalled();
  expect(reexecuteOperation).toHaveBeenCalled();
});

it('should allow providing a custom isSubscriptionOperation implementation', async () => {
  const exchangeArgs = {
    dispatchDebug: jest.fn(),
    forward: () => empty as Source<OperationResult>,
    client: {} as Client,
  };

  const isSubscriptionOperation = jest.fn(() => true);

  const forwardSubscription: SubscriptionForwarder = () => ({
    subscribe(observer) {
      observer.next(subscriptionResult);
      return { unsubscribe: jest.fn() };
    },
  });

  await pipe(
    fromValue(subscriptionOperation),
    subscriptionExchange({ forwardSubscription, isSubscriptionOperation })(
      exchangeArgs
    ),
    take(1),
    toPromise
  );

  expect(isSubscriptionOperation).toHaveBeenCalled();
});
