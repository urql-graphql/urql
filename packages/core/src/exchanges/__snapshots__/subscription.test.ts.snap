// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return response data from forwardSubscription observable 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "extensions": undefined,
  "hasNext": false,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 4,
    "kind": "subscription",
    "query": Object {
      "__key": 2088253569,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "subscribeToUser",
          },
          "operation": "subscription",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "user",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "user",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 92,
        "source": Object {
          "body": "# subscribeToUser
subscription subscribeToUser($user: String) { user(user: $user) { name } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "user": "colin",
    },
  },
}
`;
